KAFKA-12610: Implement PluginClassLoader::getResource and getResources

gharris1727 opened the pull request:
Copied from original PR #10475:

The default implementation, which is inheritted from ClassLoader, searches the
classloader tree from parent first. This causes issues when the resource is
available both on the classpath and the plugin path. Instead of attempting to
load the resource from plugin path first, the system classloader is consulted
first and loads the resource from classpath.

A testcase is added in PluginsTest to verify this behavior is fixed by this
commit.

Thanks to @ncliang for the original implementation.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ncliang committed:
KAFKA-12610: Implement PluginClassLoader::getResource

The default implementation, which is inheritted from ClassLoader, searches the
classloader tree from parent first. This causes issues when the resource is
available both on the classpath and the plugin path. Instead of attempting to
load the resource from plugin path first, the system classloader is consulted
first and loads the resource from classpath.

A testcase is added in PluginsTest to verify this behavior is fixed by this
commit.

ncliang committed:
Delete ReadVersionFromResource.java

ncliang committed:
Delete ReadVersionFromResource.java

ncliang committed:
Delete version

ncliang committed:
unused imports

ncliang committed:
Review comments

ncliang committed:
package private constructor

ncliang committed:
checkstyle

ncliang committed:
Review comment

ncliang committed:
minor: javadoc

ncliang committed:
minor: remove inaccurate comment

ncliang committed:
review comment - hide from plugin path

gharris1727 committed:
Merge branch 'trunk' into KAFKA-12610

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 committed:
Merge branch 'trunk' into KAFKA-12610

gharris1727 committed:
fixup: merge conflicts

Signed-off-by: Greg Harris <greg.harris@aiven.io>

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginClassLoader.java:
```diff
@@ -81,6 +82,17 @@ public String toString() {
+    public URL getResource(String name) {
+        Objects.requireNonNull(name);
+
+        URL url = findResource(name);
```
Probably worth adding a brief comment about how this preserves classpath loading isolation by allowing plugins to package their own resources.

There's also an interesting note in the [ClassLoader::getResources Javadocs](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/ClassLoader.html#getResources(java.lang.String)):

> When overriding this method it is recommended that an implementation ensures that any delegation is consistent with the [getResource(String)](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/ClassLoader.html#getResource(java.lang.String)) method. This should ensure that the first element returned by the Enumeration's nextElement method is the same resource that the getResource(String) method would return.


So, unless there's a good reason not to, we should also override the `getResources` method with the same child-first logic that we have in this method.

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoaderTest.java:
```diff
@@ -118,7 +119,8 @@ public void testLoadingMixOfValidAndInvalidPlugins() throws Exception {
             Path source = new File(sourceJar).toPath();
-            Files.copy(source, pluginPath.resolve(source.getFileName()));
+            Files.copy(source, pluginPath.resolve(source.getFileName()),
+                    StandardCopyOption.REPLACE_EXISTING);
```
OOC, why is this added? Shouldn't the temporary `pluginDir` folder be cleaned up after test runs?

gharris1727 replied:
Removing it doesn't seem to cause the test to fail, so I don't think it's necessary for the test to pass.

I believe you're correct that the pluginDir is cleaned up and won't have stale values from previous executions. A name collision may happen in other situations, none of which appear to be happening in this test:
* A test plugin produces an `invalid.jar` file that collides with the empty file created earlier in the test
* Two or more test plugins produce jar files with the same name

This may also have been necessary at some stage of development and then obviated earlier in the review.

I'll remove this change for clarity of the diff.

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -60,12 +60,27 @@ public Plugins(Map<String, String> props) {
+    Plugins(Map<String, String> props, ClassLoader parent) {
+        List<String> pluginLocations = WorkerConfig.pluginLocations(props);
+        delegatingLoader = newDelegatingClassLoader(pluginLocations, parent);
+        delegatingLoader.initLoaders();
```
Nit: can we reduce duplication by delegating to this method in the public constructor?
```java
    public Plugins(Map<String, String> props) {
        this(props, Plugins.class.getClassLoader());
    }
```

And if we do that, we can also get rid of the single-parameter `newDelegatingClassLoader` method.

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java:
```diff
@@ -64,56 +63,95 @@
  * the names of the different plugins directly via the exposed constants.
```
Should this be updated from `exposed constants` to `{@link TestPlugin} enum`?

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java:
```diff
@@ -64,56 +63,95 @@
 public class TestPlugins {
+    public enum TestPlugin {
+        /**
+         * Resource dir and class name of a plugin which will always throw an exception during loading
```
The "Resource dir and class name of a..." prefix for each of these is a little redundant. Can we nix that and just use, e.g., `A plugin which will always throw an exception during loading`, `A {@link org.apache.kafka.common.config.provider.ConfigProvider} which samples information about its method calls.`, etc.?

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java:
```diff
@@ -64,56 +63,95 @@
+            this.hideFromPluginPath = hideFromPluginPath;
+        }
+
+        public String getResourceDir() {
```
Nit: we don't use the `get` prefix in this code base:

```suggestion
        public String resourceDir() {
```

Also, this method is unused at the moment.

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java:
```diff
@@ -64,56 +63,95 @@
+            return resourceDir;
+        }
+
+        public String getClassName() {
```
Nit: no `get`:
```suggestion
        public String className() {
```

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java:
```diff
@@ -64,56 +63,95 @@
+         */
+        READ_VERSION_FROM_RESOURCE_V1("read-version-from-resource-v1", "test.plugins.ReadVersionFromResource"),
+        // The behavior of DelegatingClassLoader when two jars providing the same plugin are both
+        // on the plugin path is undeterministic. So, hide the v2 jar from the plugin path. This jar
```
Nit:
```suggestion
        // on the plugin path is nondeterministic. So, hide the v2 jar from the plugin path. This jar
```

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java:
```diff
@@ -64,56 +63,95 @@
+
+        private final String resourceDir;
+        private final String className;
+        private final boolean hideFromPluginPath;
```
I was a little misled by this field and the effects it'd have later on when seeing uses of `TestPlugins::pluginPath` with `READ_VERSION_FROM_RESOURCE_V2` as an argument, since based on this field that test plugin is meant to be hidden from the plugin path.

Do you think `hideByDefault` would be clearer about how plugins with this field set to `true` aren't completely hidden, but you do have to opt in to seeing them?

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java:
```diff
@@ -141,18 +179,37 @@ public static void assertAvailable() throws AssertionError {
+    /**
+     * Gets the path to a single jar containing the test plugin.
+     * @param testPlugin the unique {@link TestPlugin} enum representing the plugin
+     * @return a string representing the plugin jar filename, or null if a plugin for the
```
Any reason to return null instead of throwing an exception of some kind if the plugin cannot be found? It looks like the one place where we're currently using this method will fail with an NPE if it does end up returning null, and I can't think of any good cases where we'd want `null` here instead of an exception.

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginsTest.java:
```diff
@@ -327,6 +333,63 @@ public void newPluginsShouldConfigureWithPluginClassLoader() {
+        String pluginPath = TestPlugins.pluginPath(parentResource);
+        URLClassLoader parent = new URLClassLoader(
+                new URL[]{new File(pluginPath).toURI().toURL()});
+
```
Nit: can remove this blank line
```suggestion
```

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginsTest.java:
```diff
@@ -327,6 +333,63 @@ public void newPluginsShouldConfigureWithPluginClassLoader() {
+                TestPlugin.READ_VERSION_FROM_RESOURCE_V1,
+                TestPlugin.ALIASED_STATIC_FIELD,
+                TestPlugin.READ_VERSION_FROM_RESOURCE_V1.getClassName(),
+                "1.0.0");
+    }
+
+    @Test
+    public void pluginClassLoaderReadVersionFromResourceExistingInParentAndChild() throws Exception {
+        TestPlugins.assertAvailable();
+
+        assertClassLoaderReadsVersionFromResource(
+                TestPlugin.READ_VERSION_FROM_RESOURCE_V1,
+                TestPlugin.READ_VERSION_FROM_RESOURCE_V2,
+                TestPlugin.READ_VERSION_FROM_RESOURCE_V2.getClassName(),
+                "2.0.0");
+    }
+
+    private void assertClassLoaderReadsVersionFromResource(
+            TestPlugin parentResource, TestPlugin childResource, String className, String expectedVersion) throws MalformedURLException {
+        String pluginPath = TestPlugins.pluginPath(parentResource);
+        URLClassLoader parent = new URLClassLoader(
+                new URL[]{new File(pluginPath).toURI().toURL()});
+
+
+        // Initialize Plugins object with parent class loader in the class loader tree. This is
+        // to simulate the situation where jars exist on both system classpath and plugin path.
+        Map<String, String> pluginProps = new HashMap<>();
+        pluginProps.put(WorkerConfig.PLUGIN_PATH_CONFIG,
+                TestPlugins.pluginPath(childResource));
+        plugins = new Plugins(pluginProps, parent);
+
+        Converter converter = plugins.newPlugin(
+                className,
+                new AbstractConfig(new ConfigDef(), Collections.emptyMap()),
+                Converter.class
+        );
+        // Verify the version was read from the correct resource
+        assertEquals(expectedVersion,
+                new String(converter.fromConnectData(null, null, null)));
+    }
```
I love these test cases.

Great work @ncliang, hope you're well!

C0urante reviewed (commented):
Thanks for picking this up, @gharris1727! Mostly nits, just one serious thought w/r/t overloading `getResources` to match the recommendations for the `ClassLoader` API.

gharris1727 committed:
Review comments: implement and test getResources

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 committed:
Review comment: revert change to DelegatingClassLoaderTest

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 committed:
Review comment: remove duplicate Plugins constructor, require explicit parent for DelegatingClassLoader

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 committed:
Review comments: javadocs on TestPlugin enum

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 committed:
Review comment: no get prefix for getters

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 committed:
Review comment: rename hide flag to hideByDefault

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 committed:
Review comment: refactor method for generating plugin path to simplify error handling, remove explicit assertAvailable calls.

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 commented:
@C0urante I've refactored the `hideFromPluginPath` and `pluginPath(TestPlugin)` functionality to hopefully resolve the awkwardness of both when dealing with default/non-default plugins.

the `hideFromPluginPath` flag is now inverted to `includeByDefault`, and most plugins are marked to be included in the default plugin path. The default plugin.path is returned by `pluginPath()` with no arguments, and custom plugin.paths can be assembled by calling `pluginPath(TestPlugin...)` with all of the relevant plugins.

I've also simplified the error handling of pluginPath/pluginClasses by having them assert the integrity of the PLUGIN_JARS. Upon reflecting on the previous design, it was silly not to assert the integrity when using the variable, and force the caller to propagate errors explicitly.

These two changes should obviate the need for any error handling for a `pluginPath(singlePluginConstant)` call.

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginClassLoader.java:
```diff
@@ -81,6 +85,48 @@ public String toString() {
+                }
+                return tmp[index].nextElement();
+            }
+        };
```
We can simplify this by using the `Vector` class, which is a `List` implementation capable of providing an `Enumeration` for its elements:
```suggestion
        Vector<URL> resources = new Vector<>();
        for (Enumeration<URL> foundLocally = findResources(name); foundLocally.hasMoreElements();) {
            URL url = foundLocally.nextElement();
            if (url != null)
                resources.add(url);
        }
        // Explicitly call the parent implementation instead of super to avoid double-listing the local resources
        for (Enumeration<URL> foundByParent = getParent().getResources(name); foundByParent.hasMoreElements();) {
            URL url = foundByParent.nextElement();
            if (url != null)
                resources.add(url);
        }
        return resources.elements();
```

C0urante commented on connect/runtime/src/test/resources/test-plugins/read-version-from-resource-v1/test/plugins/ReadVersionFromResource.java:
```diff
@@ -35,22 +38,37 @@ public void configure(final Map<String, ?> configs, final boolean isKey) {
+    }
+
+    @Override
+    public byte[] fromConnectData(final String topic, final Schema schema, final Object value) {
```
We should add some Javadocs either to this class or to its `fromConnectData` and `toConnectData` methods explaining that these aren't "real" converters and the methods that they implement from that interface are just used during tests to relay information about which class loader resources were available to them at the time of invocation.

(Same goes for the "v2" variant as well)

C0urante reviewed (commented):
Thanks @gharris1727, two small comments and then this should be good to go.

gharris1727 committed:
Review comment: Vector based getResources implementation

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 committed:
Review comment: better javadocs for test plugins

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 committed:
fixup: typos in javadocs

Signed-off-by: Greg Harris <greg.harris@aiven.io>

C0urante reviewed (approved):
LGTM, thanks @gharris1727 and @ncliang!
